src/testing/slim_check/functions.lean

```diff
@@ -0,0 +1,416 @@
+/-
+Copyright (c) 2020 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+-/
+import data.list.sigma
+import testing.slim_check.sampleable
+import testing.slim_check.testable
+import tactic.pretty_cases
+
+/-!
+## `slim_check`: generators for functions
+
+This file defines `sampleable` instances for `ℤ → ℤ` functions and
+`ℤ → ℤ` injective functions.
+
+Functions are generated by creating a list of pairs and one more value
+using the list as a lookup table and resorting to the additional value
+when a value is not found in the table.
+
+Injective functions are generated by creating a list of numbers and
+a permutation of that list. The permutation insures that every input
+is mapped to a unique output. When an input is not found in the list
+the input itself is used as an output.
+-/
```

Comment:
Could you say something here about what would be necessary to define `sampleable` instances for functions between other types?

Reply:
Also, it looks like lots of list-specific stuff in this file should be moved out to `data.list`.

Reply:
A lot? I thought I moved all of that out. Maybe you you mean `nth_injective` and `nth_injective'`? Anything else?

Reply:
Maybe `perm.slice` and `slice_sizes`, too. I didn't look closely enough the first time and I thought that `apply_id` was more broadly applicable, but now I see that it is pretty specific to the `slim_check.injective_function` setup.

Reply:
I can see it's arguable for `perm.slice` . `slice_sizes` is really a part of the shrinking algorithm. I don't think it holds general interest. For `perm.slice`, I packaged it in a sigma type to make it usable in the shrinking which makes the type signature specific to this design (I'd keep adding invariants if I need them) but the proofs were all moved to `data/list/perm.lean`

Reply:
OK, I see, thanks for the explanation!